File: 백준/Silver/11279. 최대 힙/최대 힙.java
import java.io.*;
import java.util.Comparator;
import java.util.PriorityQueue;

class Main {

    static StreamTokenizer sttk = new StreamTokenizer(new InputStreamReader(System.in));

    static int readInt() throws IOException {
        sttk.nextToken();
        return (int) sttk.nval;
    }

    public static void main(String[] args) throws IOException {
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        PriorityQueue<Integer> pq = new PriorityQueue<>(Comparator.reverseOrder());
        int n = readInt();

        while (n-- > 0) {
            int k = readInt();

            if (k == 0) {
                if (!pq.isEmpty()) {
                    bw.append(Integer.toString(pq.poll()));
                } else {
                    bw.append('0');
                }
                bw.newLine();
            } else {
                pq.add(k);
            }
        }

        bw.flush();
        bw.close();
    }
}

File: 백준/Silver/11279. 최대 힙/최대 힙.kt
import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()

    val bw = System.out.bufferedWriter()
    val pq = PriorityQueue<Int>(10000, compareByDescending { it })

    repeat(n) {
        when (val x = readLine().toInt()) {
            0 -> bw.append("${pq.poll() ?: 0}\n")
            else -> pq.add(x)
        }
    }
    close()

    bw.flush()
    bw.close()
}